go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -0,0 +1,469 @@
+					klog.Warningf("failed to handle pod %v update: %v", podNew, err)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			if err := n.deletePod(pod); err != nil {
+				klog.Warningf("failed to handle pod %v deletion: %v", pod, err)
+			}
+		},
+	}, n.syncPods)
+	return err
+}
+
+func (n *NodeController) startNodeWatch(wf *factory.WatchFactory) error {
+	return houtil.StartNodeWatch(n, wf)
+}
+
+func nameToCookie(nodeName string) string {
+	hash := sha256.Sum256([]byte(nodeName))
+	return fmt.Sprintf("%02x%02x%02x%02x", hash[0], hash[1], hash[2], hash[3])
+}
+
+// hybridOverlayNodeUpdate sets up or tears down VXLAN tunnels to hybrid overlay
+// nodes in the cluster
+func (n *NodeController) hybridOverlayNodeUpdate(node *kapi.Node) error {
+	if !houtil.IsHybridOverlayNode(node) {
+		return nil
+	}
+
+	cidr, nodeIP, drMAC := getNodeDetails(node)
+	if cidr == nil || nodeIP == nil || drMAC == nil {
+		klog.V(5).Infof("cleaning up hybrid overlay tunnel to node %s", node.Name)
+		n.Delete(node)
+		return nil
+	}
+
+	klog.Infof("setting up hybrid overlay tunnel to node %s", node.Name)
+
+	// (re)add flows for the node
+	cookie := nameToCookie(node.Name)
+	drMACRaw := strings.Replace(drMAC.String(), ":", "", -1)
+
+	// Distributed Router MAC ARP responder flow; responds to ARP requests by OVN for
+	// any IP address within this node's assigned subnet and returns our hybrid overlay
+	// port's MAC address.
+	_, _, err := util.RunOVSOfctl("add-flow", extBridgeName,
+		fmt.Sprintf("cookie=0x%s,table=0,priority=100,arp,in_port=ext,arp_tpa=%s actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],mod_dl_src:%s,load:0x2->NXM_OF_ARP_OP[],move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],load:0x%s->NXM_NX_ARP_SHA[],move:NXM_OF_ARP_TPA[]->NXM_NX_REG0[],move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],move:NXM_NX_REG0[]->NXM_OF_ARP_SPA[],IN_PORT", cookie, cidr.String(), drMAC.String(), drMACRaw))
+	if err != nil {
+		return fmt.Errorf("failed to add ARP responder flow for node %q: %v", node.Name, err)
+	}
+
+	// Send all flows for the remote node's assigned subnet to that node via the VXLAN tunnel.
+	// Windows hybrid overlay implementation requires that we set the destination MAC address
+	// to the node's Distributed Router MAC.
+	_, _, err = util.RunOVSOfctl("add-flow", extBridgeName,
+		fmt.Sprintf("cookie=0x%s,table=0,priority=100,ip,nw_dst=%s,actions=load:%d->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,set_field:%s->eth_dst,output:1",
+			cookie, cidr.String(), hotypes.HybridOverlayVNI, nodeIP.String(), drMAC.String()))
+	if err != nil {
+		return fmt.Errorf("failed to add VXLAN flow for node %q: %v", node.Name, err)
+	}
+
+	return nil
+}
+
+// Add handles node additions and updates
+func (n *NodeController) Add(node *kapi.Node) {
+	var err error
+	if node.Name == n.nodeName {
+		// Retry hybrid overlay initialization if the master was
+		// slow to add the hybrid overlay logical network elements
+		err = n.ensureHybridOverlayBridge()
```

Comment:
instead of running this blindly again, how about check if it has already run on this node first, so we dont have to execute all of those ovs commands again

Reply:
@trozet pretty common pattern in the rest of ovnkube though, especially in the node addition and management port code. Hence all the --may-exist and --if-exists stuff. Also it's a kind a reconciliation hack :)

Reply:
IMO these patterns are bad. The cost of doing an ovs/ovn ctl commands in bash without a proper IDL is already poor, we shouldn't be re-issuing them again unless we have to. We can either use a variable to store that this node was already initialized or do 1 ctl cmd to check if something that should have been configured by ensure has already been done. I would prefer the former.

Reply:
@trozet so there's actually `	if n.initialized {` at the top of that function, so we already do check and not run it again.